S2-458 Add protected routes option and fix MIN_SCOPE bug.

obdeijn opened the pull request:
- Add configuration option PROTECTED to define which routes require which scopes.
- Fix bug that when MIN_SCOPE is configured as tuple (as documented), access is not granted even when required scope is in token.
- Extend unit tests

obdeijn committed:
S2-458 Add protected routes option and fix MIN_SCOPE bug.
Add configuration option PROTECTED to define which routes require which scopes.
Fix bug that when MIN_SCOPE is configured as tuple (as documented), access is not granted even when required scope is in token.

obdeijn committed:
S2-458 Small tweaks

obdeijn committed:
S2-458 Update README with new functionality.

obdeijn commented:
Added this functionality so that django projects which currently use a route based method of access configuration can easily be migrated to this middleware.

jklardie commented on authorization_django/middleware.py:
```diff
@@ -173,6 +173,13 @@ def convert_scope(scope):
+            return True
+        if '*' in protected_methods and method.upper() is not 'OPTIONS':
+            return True
+        return False
```
hier zou je het volgende van kunnen maken:

`return '*' in protected_methods and method.upper() is not 'OPTIONS'`

obdeijn replied:
Done.

jklardie commented on test_authorization_django.py:
```diff
@@ -339,13 +424,93 @@ def test_options_works_while_min_scope():
+
+    # a token with scope1 gives access via all methods
+    # to the one_scope_required route
+    request = create_request(
+       tokendata_scope1,
+       '4', 'Bearer', '/one_scope_required', 'GET'
+    )
+    response = middleware(request)
+    assert request.is_authorized_for('scope1')
+    assert response.status_code == 200
+
+    request = create_request(
+       tokendata_scope1,
+       '4', 'Bearer', '/one_scope_required', 'POST'
+    )
+    response = middleware(request)
+    assert request.is_authorized_for('scope1')
+    assert response.status_code == 200
```
Dit zou ik liever in een loop zien:
```
for method in ('GET', 'POST'):
    request = create_request(
       tokendata_scope1,
       '4', 'Bearer', '/one_scope_required', method
    )
    response = middleware(request)
    assert request.is_authorized_for('scope1')
    assert response.status_code == 200
```

obdeijn replied:
Je hebt gelijk, ik het het aangepast.

jklardie reviewed (approved):
Ziet er in het algemeen erg goed uit. Twee kleine comments, maar niks inhoudelijks.

obdeijn committed:
S2-458 Small refactor, Apply feedback from jklardie

jklardie reviewed (approved):
LGTM

tcoenen commented on README.md:
````diff
@@ -38,24 +38,59 @@ your ``settings.py`` in the ``DATAPUNT_AUTHZ`` dictionary.
+```
+
+#### Define protected routes
+With the PROTECTED setting you can define routes that require certain scopes for access. A distinction can me made between HTTP methods. An exception is made for the OPTIONS method, which is always allowed.
````
Spelling: "can me made" -> "can be made"

obdeijn replied:
Check

tcoenen commented on README.md:
````diff
@@ -38,24 +38,59 @@ your ``settings.py`` in the ``DATAPUNT_AUTHZ`` dictionary.
 `authz_level`:
 
 ```
 import authorization_django
````
Unused import.

obdeijn replied:
Ah yes, removed it.

tcoenen reviewed (commented):
lgtm

Two small things in the README

tcoenen reviewed (approved):
akkoord

obdeijn committed:
S2-458 Do some more config checks
- Raise exception when path in PROTECTED conflicts with path in white
- Raise exception when resource is in PROTECTED but no scopes have been set as required
- Add some more specific config exception classes
- Add some more unit tests
- Update README
